cpmpy/transformations/linearize.py

```diff
@@ -5,190 +5,187 @@
-                return [Comparison(cpm_expr.name, Operator("wsum", _wsum_make(lhs)), rhs)]
-            return [cpm_expr]
+            # BV -> LinExpr
+            if isinstance(cond, _BoolVarImpl):
+                lin_sub = linearize_constraint(sub_expr, supported=supported, reified=True)
+                return [cond.implies(lin) for lin in lin_sub]
 
+    # comparisons
+    if isinstance(cpm_expr, Comparison):
         lhs, rhs = cpm_expr.args
-        var, constraints = get_or_make_var(lhs.args[0] * lhs.args[1])
-        # decompose long multiplication in several multiplications with two args
-        for arg in lhs.args[2:]:
-            var, cons = get_or_make_var(var * arg)
-            constraints += cons
 
-        return constraints + [Comparison(cpm_expr.name, var, rhs)]
+        if isinstance(lhs, _NumVarImpl) and isinstance(rhs, _NumVarImpl):
+            # bring numvar to lhs
+            lhs = 1 * lhs + -1 * rhs
+            rhs = 0
+
+        # linearize unsupported operators
+        elif isinstance(lhs, Operator) and lhs.name not in supported: # TODO: add mul, (abs?), (mod?), (pow?)
+
+            if lhs.name == "sub":
+                # convert to wsum
+                lhs = sum([1 * lhs.args[0] + -1 * lhs.args[1]])
+
+            elif lhs.name == "mul" and is_num(lhs.args[0]):
+                lhs = Operator("wsum",[[lhs.args[0]], [lhs.args[1]]])
+            else:
+                raise TransformationNotImplementedError(f"lhs of constraint {cpm_expr} cannot be linearized, should be any of {supported | set(['sub'])} but is {lhs}. Please report on github")
+
+        elif isinstance(lhs, GlobalConstraint) and lhs.name not in supported:
+            raise ValueError("Linearization of `lhs` not supported, run `cpmpy.transformations.decompose_global.decompose_global() first")
+
+        if is_num(lhs) or (isinstance(lhs, Operator) and lhs.name in {"sum","wsum"}):
+            # bring all vars to lhs
+            if isinstance(rhs, _NumVarImpl):
+                if isinstance(lhs, Operator) and lhs.name == "sum":
+                    lhs, rhs = sum([1 * a for a in lhs.args]+[-1 * rhs]), 0
+                else:
+                    lhs, rhs = lhs + -1*rhs, 0
+
+            # bring all const to rhs
+            if lhs.name == "sum":
+                new_args = []
+                for i, arg in enumerate(lhs.args):
+                    if is_num(arg):
+                        rhs -= arg
+                    else:
+                        new_args.append(arg)
+                lhs = Operator("sum", new_args)
+
+            elif lhs.name == "wsum":
+                new_weights, new_args = [],[]
+                for i, (w, arg) in enumerate(zip(*lhs.args)):
+                    if is_num(arg):
+                        rhs -= w * arg
+                    else:
+                        new_weights.append(w)
+                        new_args.append(arg)
+                lhs = Operator("wsum",[new_weights, new_args])
+
```

Comment:
How about the case where `is_num(lhs)` ?

Reply:
All variables are on the lhs at this point, so if `is_num(lhs)` is true here, we have a simple Python comparison which will automatically be evaluated. I'll add an assert here to make it more explicit